packages/flutter/lib/src/cupertino/text_selection_toolbar.dart

```diff
@@ -510,21 +527,35 @@ class _CupertinoTextSelectionToolbarContentState extends State<_CupertinoTextSel
   Widget build(BuildContext context) {
     return widget.toolbarBuilder(context, widget.anchor, widget.isAbove, FadeTransition(
       opacity: _controller,
-      child: _CupertinoTextSelectionToolbarItems(
-        page: _page,
-        backButton: CupertinoTextSelectionToolbarButton.text(
-          onPressed: _handlePreviousPage,
-          text: '◀',
-        ),
-        dividerWidth: 1.0 / MediaQuery.devicePixelRatioOf(context),
-        nextButton: CupertinoTextSelectionToolbarButton.text(
-          onPressed: _handleNextPage,
-          text: '▶',
-        ),
-        nextButtonDisabled: const CupertinoTextSelectionToolbarButton.text(
-          text: '▶',
+      child: AnimatedSize(
+        duration: _kToolbarTransitionDuration,
+        curve: Curves.decelerate,
+        child: GestureDetector(
+          onHorizontalDragEnd: _onHorizontalDragEnd,
+          child: _CupertinoTextSelectionToolbarItems(
+            key: _toolbarItemsKey,
+            page: _page,
+            backButton: CupertinoTextSelectionToolbarButton(
+              onPressed: _handlePreviousPage,
+              child: Icon(
+                CupertinoIcons.chevron_left,
```

Comment:
Ah, looks like for 2. we may also be able to override mirroring in RTL by simply supplying `textDirection` to `Icon`>

Reply:
For 1, are you worried about increasing app size?  I see that CupertinoIcons is used elsewhere in Cupertino, like [in CupertinoTextField](https://github.com/flutter/flutter/blob/ca495da92a18a4b8bcd7f7953a08cbff73153832/packages/flutter/lib/src/cupertino/text_field.dart#L1195).  @MitchellGoodwin do you know if there's any concern with using CupertinoIcons like this?

For 2, good call.  @luccasclezar should try it out and make sure it looks right in RTL.

Reply:
Yes, CupertinoIcons are used throughout, including CupertinoApp itself, so I do not think it will increase the size at all. Only concerns with the package is that it's somewhat out of date, but the chevron icons aren't an issue.

Reply:
About 2, I actually noticed this, but I'm using `CupertinoIcons.chevron_left` (and right), **not** `CupertinoIcons.left_chevron`. The only difference between the two if you look at the source code is that `chevron_left` does not have `matchTextDirection` enabled.

https://github.com/luccasclezar/flutter/blob/a36b1f9912de66a2588a7e9bababc6133b29fc31/packages/flutter/lib/src/cupertino/icons.dart#L1767

Regardless of the theory, I ran Flutter Gallery, enabled "Force RTL" and added a bunch of "Select All" items to the toolbar, this is the result:

![Screenshot_20230606_211838](https://github.com/flutter/flutter/assets/12024080/bb754aa3-355b-489c-b1ad-94eaec56d087)

Reply:
@justinmc @MitchellGoodwin wouldn't this require adding the `cupertino_icons` dependency? Note that the Cupertino toolbar is also used for the regular `TextField` e.g. in a `MaterialApp`.

Reply:
The Cupertino icons is a static internal package that was generated based on an older version of the outside package `cupertino_icons`. So Flutter comes with them already ready to go, and we need to manually update the Cupertino icons to catch changes to `cupertino_icons` so we aren't in danger of breaking changes from that package.

Reply:
@MitchellGoodwin What about for apps that don't use Cupertino at all, will the icons be included in the bundle?  If so, then I believe that someone that isn't using Cupertino at all, but does use a TextField, will now have the Cupertino icons included due to this PR's change.

I wouldn't be surprised if this PR doesn't actually impact this in practice, but it would be nice to fully understand it.  Maybe we need to rope someone in that is an expert here.

Reply:
I see. In that case, I *believe* that because it's an internal package the unused icons will be tree shaken out. For outside font packages, if you import them it will tree shake successfully as long as you use at least [one constant from that package](https://github.com/flutter/flutter/issues/63920).

I'm unsure of what the differences would be here. It's possible that the whole icon collection would be included unnecessarily in this case, but I think only the required ones would be.

Reply:
Thanks for investigating.  Hopefully the unused icons do get tree shaken out, but if not, then it's something to solve separately from this PR.  I'm sure there are many Flutter iOS apps that use TextField and want these toolbar button updates.

I think we're fine to move ahead with this PR with @tgucio's approval.

Reply:
Here's what happens in a `MaterialApp` using a regular `TextField` on iOS if `cupertino_icons` isn't in the app dependencies (just swapped labels in current master with these icons and made strings longer):

<img width="508" alt="toolbar" src="https://github.com/flutter/flutter/assets/72562119/a1002bee-6a29-4bc1-b97c-4ef3b5e47b47">

For icon tree shaked size, this example results in:
```
% ls -al build/app/intermediates/assets/release/flutter_assets/packages/cupertino_icons/assets/CupertinoIcons.ttf
-rw-r--r--@ 1 tgucio  staff  976 Jun  8 16:36 build/app/intermediates/assets/release/flutter_assets/packages/cupertino_icons/assets/CupertinoIcons.ttf
```

So size wise it's tiny but I'm still not sure about requiring the `cupertino_icons` dependency.

Maybe we can revert this part and use the former `CupertinoTextSelectionToolbarButton.text` chars - and open another PR to continue this discussion?

Reply:
I managed to remove the CupertinoIcons by using a `CustomPainter` to draw the chevrons.

The tests are failing because one, I didn't want to change them before asking what you think about this, and two, I want to know your opinions on what would be the best way to test the painter (probably make `_CupertinoChevronPainter` public with `@visibleForTesting`?)

Reply:
Good stuff! If @justinmc and @MitchellGoodwin don't have objections on this being the best approach to avoid the `cupertino_icons` dependency I'd say it's good to go.

For checking what a `CustomPainter` paints, there's a matcher `paints` with many examples e.g. in flutter/test/cupertino/context_menu_action_test.dart:
```
    await tester.pumpWidget(getApp());
    expect(find.byType(CupertinoContextMenuAction),
        paints..rect(color: kBackgroundColor.color));
```

Negative test example:
```
    expect(
      tester.firstRenderObject<RenderBox>(find.byType(CupertinoRadio<int>)),
      isNot(paints..path())
    );
```

Reply:
Good catch!  I didn't realize cupertino_icons would have to be explicitly included.  I'm on board with the custom painter solution :+1:

Reply:
@tgucio Thank you for pointing out how to test the CustomPainter, I added some tests to verify that the chevrons are pointing the correct way depending on `_CupertinoChevronPainter.isLeft` 😁

But there's another problem: in the current tests, there are lines checking if the back/next buttons are being shown, but to know this I would need a way to find a `CustomPainter` with a `_CupertinoChevronPainter` as painter and access its `isLeft` field.

The icons were found like this in the previous implementation:

```dart
  Finder findOverflowNextButton() => find.byIcon(CupertinoIcons.chevron_right);
  Finder findOverflowBackButton() => find.byIcon(CupertinoIcons.chevron_left);
```

And I could keep the same structure if `_CupertinoChevronPainter` was public, something like this would work:

```dart
  Finder findOverflowNextButton() => find.byWidgetPredicate((Widget widget) =>
    widget is CustomPaint &&
    widget.painter is CupertinoChevronPainter &&
    !((widget.painter! as CupertinoChevronPainter).isLeft,
  );
  Finder findOverflowNextButton() => find.byWidgetPredicate((Widget widget) =>
    widget is CustomPaint &&
    widget.painter is CupertinoChevronPainter &&
    ((widget.painter! as CupertinoChevronPainter).isLeft,
  );
```

But probably declaring such a specific class as public is not a good idea.

The only way I found of keeping it private but still verify back/next buttons is by checking the `runtimeType` of `CustomPainter.painter` and then, if it is `_CupertinoChevronPainter`, cast it to `dynamic` and get `isLeft`:

```dart
  Finder findOverflowNextButton() => find.byWidgetPredicate((Widget widget) =>
    widget is CustomPaint &&
    '${widget.painter?.runtimeType}' == '_CupertinoChevronPainter' &&
    !((widget.painter! as dynamic).isLeft as bool),
  );
  Finder findOverflowBackButton() => find.byWidgetPredicate((Widget widget) =>
    widget is CustomPaint &&
    '${widget.painter?.runtimeType}' == '_CupertinoChevronPainter' &&
    ((widget.painter! as dynamic).isLeft as bool),
  );
```

And even then, I don't why but I had to change all `tester.tap(findOverflowNextButton())` to `tester.tapAt(tester.getCenter(findOverflowNextButton()))` (same for the back button).

So, there's probably a better way to do this? 😅

Reply:
Actually, I didn't manage to test that the chevrons are pointing the correct side. The patterns are like this:

```dart
PaintPattern overflowNextPaintPattern() => paints..path(includes: <Offset>[const Offset(7.5, 5)]);
PaintPattern overflowBackPaintPattern() => paints..path(includes: <Offset>[const Offset(2.5, 5)]);
```

But even if the offsets are swapped between the two, the tests pass anyway. I think this is because `path.contains(offset)` ignores strokes and just checks if the `offset` is inside its canvas rect if there's no `fill`, and as both canvas are 10x10, both offsets always pass.

Reply:
Hey @justinmc @tgucio, I think you missed the notifications for my last two comments, but I would really appreciate your input here :)

Reply:
Ah sorry, how about something like this?

```dart
   ...

    Finder findOverflowButton() => find.byWidgetPredicate((Widget widget) =>
      widget is CustomPaint &&
      '${widget.painter?.runtimeType}' == '_CupertinoChevronPainter',
    );

    expect(findOverflowButton(), findsOneWidget);
    expect(
      findOverflowButton().last,
      paints
          ..path(
            includes: const <Offset>[
              Offset(7.5, 8.0),
              Offset(2.5, 13.0),
            ],
          ),
    );

    await tester.tap(findOverflowButton());
    await tester.pumpAndSettle();

    expect(findOverflowButton(), findsNWidgets(2));
    expect(
      findOverflowButton().first,
      paints
          ..path(
            includes: const <Offset>[
              Offset(2.5, 8.0),
              Offset(7.5, 13.0),
            ],
          ),
    );
    expect(
      findOverflowButton().last,
      paints
          ..path(
            includes: const <Offset>[
              Offset(7.5, 8.0),
              Offset(2.5, 13.0),
            ],
          ),
    );

```

Reply:
@justinmc Thank you for the suggestion :)

I don't know if `8.0` and `13.0` are the right numbers for `Offset.dy`, the chevron painter path is like this: `[(2.5, 0), (7.5, 5), (2.5, 10)]` when pointing right and `[(7.5, 0), (2.5, 5), (7.5, 10)]` when pointing left. The thing is, even tweaking this numbers like I did [here](https://github.com/flutter/flutter/pull/127757#discussion_r1231075344), it doesn't work.

Basically `paints..path()` only considers the path's `fill`, not its `stroke`, and as both the left and right chevrons have the same square area, there's no way to tell them apart with `painter..path()`.

Reply:
Found a way to properly test everything! I will update the tests tomorrow and commit, then you can take a look 😁

Reply:
I just committed, but the tl;dr is: `paints..path()` tests just the `fill`, but `paints..line()` can test the exact points that a line was drawn on, so I changed the `_CupertinoChevronPainter` to use two `canvas.drawLine`s instead of a single `canvas.drawPath`.

That made possible to test if the chevrons point to the correct side, but to actually find the chevrons I created two more private classes, one for each chevron, and now it's just a matter of comparing the `runtimeType` of the painter with either `_LeftCupertinoChevronPainter` or `_RightCupertinoChevronPainter`.

Reply:
Ha, I don't know why I didn't think of splitting it into two different painter classes, good call.  I'll give this another review.